[2.next] Make test suites compatible with PHPUnit 4.8

chinpei215 opened the pull request:
- Made test suites compatible with PHPUnit 4.8
- Fixed XML output of HelpFormatter (Backport from #10339. The related test cases don't pass with PHPUnit 4.8. I noticed this issue when I was working to make CakePHP 2.x compatible with PHPUnit 4.8 for my project. Then I checked 3.x branch and noticed that the related test cases had been fixed in a wrong way. That was why I reported #10339).
- Added PHPUnit 4.8 to test matrix
- ~~Added CakeFinalizer class. PHPUnit 4.8 marks test cases as 'risky', if the output buffering level is different before and after the test. CakeFinalizer allows us to adjust the output buffering level, when we want to test some function which changes the output buffering level and throws an exception.~~
- Fixed several risky tests.
    - Fixed CakeTestCase::run() to restore original output buffering level
    - Fixed an undefined variable warning in missing_controller.ctp
    - Fixed ViewTest (backport of 92bd86274b64c31ef6034caa74bb168be041b5e0)
    - Fixed ThemeViewTest

I have not changed `composer.json` as I think that using PHPUnit 4.8 is optional. So, for now, only a test is running with PHPUnit 4.8. However, It may be arguable. I am fine to change `composer.json` and `.travis.yml` to let all test cases run with PHPUnit 4.8.

Refs: #10398

chinpei215 committed:
Make test suites compatible with PHPUnit 4.8

chinpei215 committed:
Fix XML output of HelpFormatter

Backport from #10339

chinpei215 committed:
Add PHPUnit 4.8 to test matrix

chinpei215 commented:
I may have to merge 2.x into 2.next and rebase this pull request, to fix an apcu issue: https://github.com/cakephp/cakephp/commit/173f927ef30f2032d14f383e58da4bface5bf130

May I merge 2.x into 2.next?

dereuromark commented:
Yeah, sure!

markstory commented on lib/Cake/Test/Case/Utility/CakeFinalizerTest.php:
```diff
@@ -0,0 +1,50 @@
+ * @return void
+ */
+	public function testCallable() {
+		$mock = $this->getMock('stdClass', array('finalize'));
```
We should probably start using the un-deprecated `getMockBuilder()` method for new tests.

markstory commented on lib/Cake/Test/Case/View/ThemeViewTest.php:
```diff
@@ -232,11 +228,7 @@ public function testMissingLayout() {
-		ob_start();
 		$View->getLayoutFileName();
-		$expected = ob_get_clean();
-		$this->assertRegExp("/Missing Layout/", $expected);
```
Should the exception message have an assertion added?

chinpei215 commented:
Thank you for your comments. Probably, I'll take a look tomorrow.

chinpei215 commented:
Rebased. But I am re-thinking about new CakeFinalizer class. First I thought it was nice idea, but I changed my mind. Now it is being used only in ViewHelperTest. I will remove CakeFinalizer and fix related test cases instead.

chinpei215 committed:
Fix risky tests

- Fix CakeTestCase::run() to restore original output buffering level
- Fix an undefined variable warning in missing_controller.ctp
- Fix ViewTest (backport of 92bd86274b64c31ef6034caa74bb168be041b5e0)
- Fix ThemeViewTest

chinpei215 commented:
Removed CakeFinalizer and fixed related risky tests. All tests has been passed with both PHPUnit 3.7 and 4.8. By the way, should we bump the PHPUnit version to 4.8? I am fine to change `composer.json` and `.travis.yml`. Then the version constraint in `composer.json` will be `>=3.7 <5.0` for the backward compatibility.

markstory commented:
Adding a more flexible version constraint on phpunit sounds like a good idea.
